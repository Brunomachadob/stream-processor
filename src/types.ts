export type PromiseResolve<T> = (result: T) => void;
export type PromiseReject = (e: Error) => void;

export type Mapper<T, U> = (chunk: T, index: number, encoding: string) => U | Promise<U>;
export type Filter<T> = (chunk: T, encoding?: string) => boolean | Promise<boolean>;
export type Reducer<T, U> = (acc: U, chunk: T) => U;
export type KeyMapper<T, K> = (chunk: T) => K;
