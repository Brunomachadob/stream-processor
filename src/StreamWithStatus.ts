import { Transform, TransformCallback } from "stream";

export class StreamWithStatus extends Transform {
    private receivedData: boolean = false;
    private isFinished: boolean = false;

    constructor() {
        super({
            objectMode: true,
            transform: (chunk: any, _: string, callback: TransformCallback) => {
                this.receivedData = true;

                callback(undefined, chunk);
            },
        });

        this.once("end", () => this.isFinished = true);
    }

    public wasEmpty(): boolean {
        if (!this.isFinished) {
            throw new Error("It's not possible to check if stream was empty before it finished transferring.");
        }

        return !this.receivedData;
    }
}
