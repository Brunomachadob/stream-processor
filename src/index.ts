export { StreamProcessor } from "./StreamProcessor";
export { StreamWithStatus } from "./StreamWithStatus";

export { Mapper, Filter, Reducer, KeyMapper } from "./types";
