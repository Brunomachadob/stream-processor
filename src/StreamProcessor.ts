// tslint:disable:max-file-line-count
import { Writable, Transform, TransformCallback, Stream, PassThrough } from "stream";

import { PromiseResolve, PromiseReject, Mapper, Filter, Reducer, KeyMapper } from "./types";

import { StreamWithStatus } from "./StreamWithStatus";

export class StreamProcessor<T> {
    private static pipeStream<T extends Writable = Writable>(src: Stream, dest: T): T {
        src.on("error", (e: Error) => dest.emit("error", e));

        return src.pipe(dest);
    }

    private readonly head: Writable;
    private tail: Writable;

    constructor() {
        const pass: PassThrough = new PassThrough({ objectMode: true });
        this.head = pass;
        this.tail = pass;
    }

    public start(source: Stream): StreamWithStatus {
        StreamProcessor.pipeStream(source, this.head);

        return StreamProcessor.pipeStream(this.tail, new StreamWithStatus());
    }

    public startMultiple(source: Stream, quantity: number): StreamWithStatus[] {
        const result: Stream = this.start(source);

        return Array.from({ length: quantity }, () => {
            return StreamProcessor.pipeStream(result as Writable, new StreamWithStatus());
        });
    }

    public map<U>(mapper: Mapper<T, U>): StreamProcessor<U> {
        let index: number = 0;

        const transformer: Transform = new Transform({
            objectMode: true,
            transform: async (chunk: any, encoding: string, callback: TransformCallback) => {
                try {
                    const mapped: U = await mapper(chunk, index++, encoding);
                    callback(undefined, mapped);
                } catch (e) {
                    return callback(e);
                }
            },
        });

        return this.pipe(transformer) as any;
    }

    public flatMap<U>(mapper: Mapper<T, U[]> = (chunk: T) => (chunk as any)): StreamProcessor<U> {
        let index: number = 0;

        const transformer: Transform = new Transform({
            objectMode: true,
            async transform(chunk: T, encoding: string, callback: TransformCallback): Promise<void> {
                try {
                    const mapped: U[] = await mapper(chunk, index++, encoding);
                    mapped.forEach((item: U) => {
                        this.emit("data", item);
                    });

                    callback();
                } catch (e) {
                    callback(e);
                }
            },
        });

        return this.pipe(transformer) as any;
    }

    public filter(filterFn: Filter<T>): StreamProcessor<T> {
        const transformer: Transform = new Transform({
            objectMode: true,
            transform: async (chunk: any, encoding: string, callback: TransformCallback) => {
                try {
                    const accept: boolean = await filterFn(chunk, encoding);
                    callback(undefined, accept ? chunk : undefined);
                } catch (e) {
                    return callback(e);
                }
            },
        });

        return this.pipe(transformer);
    }

    public reduce<U>(reducer: Reducer<T, U>, initial?: U): StreamProcessor<U> {
        let acc: U | undefined = initial;

        const transformer: Transform = new Transform({
            objectMode: true,
            transform(chunk: T, _: string, callback: TransformCallback): void {
                try {
                    if (acc !== undefined) {
                        acc = reducer(acc, chunk);
                    } else {
                        acc = chunk as any;
                    }

                    callback();
                } catch (e) {
                    callback(e);
                }
            },
            flush(callback: TransformCallback): void {
                callback(undefined, acc);
            },
        });

        return this.pipe(transformer) as any;
    }

    public async collect(source: Stream): Promise<T[]> {
        const arr: T[] = [];

        return this.collectFromStream(source, (data: T) => {
            arr.push(data);
        }, () => arr);
    }

    public async groupBy<K>(source: Stream, keyMapper: KeyMapper<T, K>): Promise<Map<K, T[]>> {
        const groups: Map<K, T[]> = new Map();

        return this.collectFromStream(source, (data: T) => {
            const key: K = keyMapper(data);
            let list: T[] | undefined = groups.get(key);

            if (!list) {
                list = [];
                groups.set(key, list);
            }

            list.push(data);
        }, () => groups);
    }

    private pipe(dest: Writable): StreamProcessor<T> {
        this.tail = StreamProcessor.pipeStream(this.tail, dest);

        return this;
    }
    private async collectFromStream(
        source: Stream,
        onData: (v: T) => void,
        getResult: () => any,
    ): Promise<any> {
        return new Promise<T[]>((resolve: PromiseResolve<T[]>, reject: PromiseReject) => {
            this.start(source)
                .on("data", (data: T) => {
                    try {
                        onData(data);
                    } catch (e) {
                        reject(e);
                    }
                })
                .on("end", () => resolve(getResult()))
                .on("error", reject);
        });
    }
}
