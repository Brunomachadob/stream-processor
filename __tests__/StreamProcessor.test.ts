// Is required by inversify
import { Readable } from "stream";

import { StreamProcessor, StreamWithStatus } from "../src";
import { delay, randomDelay } from "./util/delay";
import { buildStream } from "./util/buildStream";

type SimpleItem = {
    itemId: number,
    itemName: string,
}

type SimpleModel = {
    id: number,
    name: string,
    age: number,
    country: string,
    items: SimpleItem[],
};

const SIMPLE_MODEL_ARR: SimpleModel[] = [
    { id: 1, name: "Test1", age: 18, country: "DE", items: [{ itemId: 1, itemName: "Item1" }, { itemId: 2, itemName: "Item2" }] },
    { id: 2, name: "Test2", age: 20, country: "DE" ,items: [{ itemId: 3, itemName: "Item3" }, { itemId: 4, itemName: "Item4" }] },
    { id: 3, name: "Test3", age: 26, country: "BR", items: [{ itemId: 5, itemName: "Item5" }, { itemId: 6, itemName: "Item6" }] },
    { id: 4, name: "Test4", age: 23, country: "BR", items: [{ itemId: 7, itemName: "Item7" }, { itemId: 8, itemName: "Item8" }] },
];

describe("StreamProcessor", () => {
    it("StreamProcessor.collect", async () => {
        const result: number[] = await new StreamProcessor<number>()
            .collect(buildStream([ 1, 2, 3 ]));

        expect(result).toEqual([ 1, 2, 3 ]);
    });

    it("StreamProcessor.map", async () => {
        const result: number[] = await new StreamProcessor<number>()
            .map((n: number) => n * 2)
            .collect(buildStream([ 1, 2, 3 ]));

        expect(result).toEqual([ 2, 4, 6 ]);
    });

    it("StreamProcessor.map with error", async () => {
        const check: () => Promise<number[]> = async () => {
            return new StreamProcessor<number>().map((n: number) => {
                if (n > 2) {
                    throw new Error(`Throwing error because it's greather than 2, got ${n}`);
                }

                return n * 2;
            })
            .collect(buildStream([ 1, 2, 3 ]));
        };

        await expect(check()).rejects.toThrow("Throwing error because it's greather than 2, got 3");
    });

    it("StreamProcessor.flatMap", async () => {

        const result: SimpleItem[] = await new StreamProcessor<SimpleModel>()
            .flatMap((model: SimpleModel) => delay(randomDelay(100, 200), model.items))
            .collect(buildStream(SIMPLE_MODEL_ARR));

        expect(result).toMatchSnapshot();
    });

    it("StreamProcessor.flatMap with error", async () => {
        const check: () => Promise<SimpleItem[]> = async () => {
            return new StreamProcessor<SimpleModel>()
            .flatMap((model: SimpleModel) => {
                if (model.items.length > 1) {
                    throw new Error(`Throwing error because it's greather than 1, got ${model.items.length}`);
                }

                return model.items;
            })
            .collect(buildStream(SIMPLE_MODEL_ARR));
        };

        await expect(check()).rejects.toThrow("Throwing error because it's greather than 1, got 2");
    });

    it("StreamProcessor.filter", async () => {
        const result: number[] = await new StreamProcessor<number>()
            .filter((n: number) => n > 10)
            .collect(buildStream([ 10, 20, 30 ]));

        expect(result).toEqual([ 20, 30 ]);
    });

    it("StreamProcessor.filter with error", async () => {
        const check: () => Promise<number[]> = async () => {
            return new StreamProcessor<number>().filter((n: number): boolean => {
                if (n > 2) {
                    throw new Error(`Throwing error because it's greather than 2, got ${n}`);
                }

                return true;
            })
            .collect(buildStream([ 1, 2, 3 ]));
        };

        await expect(check()).rejects.toThrow("Throwing error because it's greather than 2, got 3");
    });

    it("StreamProcessor.reduce", async () => {
        const input: Readable = buildStream([ 10, 20, 30 ]);
        const reducer = (acc: number, n: number): number => acc + n;

        const [ result ]: number[] = await new StreamProcessor<number>()
            .reduce(reducer).collect(input);

        expect(result).toEqual(60);
    });

    it("StreamProcessor.reduce with different types", async () => {
        const reducer = (acc: SimpleItem[], model: SimpleModel): SimpleItem[] => [ ...acc, ...model.items ];
        const flatMapper = (items: SimpleItem[]): string[] => items.map((item: SimpleItem) => item.itemName);

        const result: string[] = await new StreamProcessor<SimpleModel>()
            .reduce(reducer, [])
            .flatMap(flatMapper)
            .collect(buildStream(SIMPLE_MODEL_ARR));

        expect(result).toMatchSnapshot();
    });

    it("StreamProcessor.reduce with error", async () => {
        const input: Readable = buildStream([ 1, 2, 3 ]);
        const reducer = (acc: number, n: number): number => {
            if (n > 0) {
                throw new Error(`Throwing error because it's greather than 0, got ${n}`);
            }

            return acc + n;
        };

        const check: () => Promise<number[]> = async () => {
            return new StreamProcessor<number>().reduce(reducer).collect(input);
        };

        await expect(check()).rejects.toThrow("Throwing error because it's greather than 0, got 2");
    });

    it("StreamProcessor.groupBy", async () => {
        const input: Readable = buildStream(SIMPLE_MODEL_ARR);
        const keyMapper = (model: SimpleModel): string => model.country;

        const result: Map<string, SimpleModel[]> = await new StreamProcessor<SimpleModel>()
            .groupBy(input, keyMapper);

        expect(result).toMatchSnapshot();
    });

    it("StreamProcessor complex operations", async () => {
        const input: Readable = buildStream(SIMPLE_MODEL_ARR);
        const reducer = (acc: number, n: number): number => acc + n;

        const [ result ]: number[] = await new StreamProcessor<SimpleModel>()
            .filter((model: SimpleModel) => model.country === "BR")
            .map((model: SimpleModel) => delay(200, model.age))
            .reduce(reducer)
            .collect(input);

        expect(result).toEqual(49);
    });

    it("multiple StreamProcessors sharing one source", async () => {
        const input: Readable = buildStream(SIMPLE_MODEL_ARR);
        const reducer = (acc: number, n: number): number => acc + n;

        const brazilianStream: StreamProcessor<number> = new StreamProcessor<SimpleModel>()
            .filter((model: SimpleModel) => model.country === "BR")
            .map((model: SimpleModel) => delay(200, model.age));

        const germanStream: StreamProcessor<number> = new StreamProcessor<SimpleModel>()
            .filter((model: SimpleModel) => model.country === "DE")
            .map((model: SimpleModel) => delay(500, model.age));

        const [ [ brazilianAges ], [ germanAges ] ]: [number[], number[]] = await Promise.all([
            brazilianStream.reduce(reducer).collect(input),
            germanStream.reduce(reducer).collect(input),
         ]);

        expect(brazilianAges).toEqual(49);
        expect(germanAges).toEqual(38);
    });

    it("it should throw an error if ask 'wasEmpty' before done", () => {
        const stream: StreamWithStatus = new StreamProcessor<number>()
            .start(buildStream([1, 2, 3]));

        expect(() => stream.wasEmpty()).toThrow("It's not possible to check if stream was empty before it finished transferring.")
    });

    it("'wasEmpty' should return false if items were transferred", async () => {
        const stream: StreamWithStatus = new StreamProcessor<number>()
            .start(buildStream([1, 2, 3]));

        stream.on("data", () => {}); // To trigger the stram start

        await new Promise((resolve: () => any) => {
            stream.on("end", () => resolve());
        });

        expect(stream.wasEmpty()).toEqual(false)
    });

    it("'wasEmpty' should return true if no items were transferred", async () => {
        const stream: StreamWithStatus = new StreamProcessor<number>()
            .start(buildStream([]));

        stream.on("data", () => {}); // To trigger the stram start

        await new Promise((resolve: () => any) => {
            stream.on("end", () => resolve());
        });

        expect(stream.wasEmpty()).toEqual(true)
    });
});
