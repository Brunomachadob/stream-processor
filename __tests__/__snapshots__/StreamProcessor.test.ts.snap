// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StreamProcessor StreamProcessor.flatMap 1`] = `
Array [
  Object {
    "itemId": 1,
    "itemName": "Item1",
  },
  Object {
    "itemId": 2,
    "itemName": "Item2",
  },
  Object {
    "itemId": 3,
    "itemName": "Item3",
  },
  Object {
    "itemId": 4,
    "itemName": "Item4",
  },
  Object {
    "itemId": 5,
    "itemName": "Item5",
  },
  Object {
    "itemId": 6,
    "itemName": "Item6",
  },
  Object {
    "itemId": 7,
    "itemName": "Item7",
  },
  Object {
    "itemId": 8,
    "itemName": "Item8",
  },
]
`;

exports[`StreamProcessor StreamProcessor.groupBy 1`] = `
Map {
  "DE" => Array [
    Object {
      "age": 18,
      "country": "DE",
      "id": 1,
      "items": Array [
        Object {
          "itemId": 1,
          "itemName": "Item1",
        },
        Object {
          "itemId": 2,
          "itemName": "Item2",
        },
      ],
      "name": "Test1",
    },
    Object {
      "age": 20,
      "country": "DE",
      "id": 2,
      "items": Array [
        Object {
          "itemId": 3,
          "itemName": "Item3",
        },
        Object {
          "itemId": 4,
          "itemName": "Item4",
        },
      ],
      "name": "Test2",
    },
  ],
  "BR" => Array [
    Object {
      "age": 26,
      "country": "BR",
      "id": 3,
      "items": Array [
        Object {
          "itemId": 5,
          "itemName": "Item5",
        },
        Object {
          "itemId": 6,
          "itemName": "Item6",
        },
      ],
      "name": "Test3",
    },
    Object {
      "age": 23,
      "country": "BR",
      "id": 4,
      "items": Array [
        Object {
          "itemId": 7,
          "itemName": "Item7",
        },
        Object {
          "itemId": 8,
          "itemName": "Item8",
        },
      ],
      "name": "Test4",
    },
  ],
}
`;

exports[`StreamProcessor StreamProcessor.reduce with different types 1`] = `
Array [
  "Item1",
  "Item2",
  "Item3",
  "Item4",
  "Item5",
  "Item6",
  "Item7",
  "Item8",
]
`;
